src/middlewares/languageMiddleware.ts

```diff
@@ -0,0 +1,169 @@
+import Negotiator from "negotiator";
+import { headers } from "next/headers";
+import { NextRequest, NextResponse } from "next/server";
+
+import { client } from "studio/lib/client";
+import { SlugTranslations } from "studio/lib/interfaces/slugTranslations";
+import { LanguageObject } from "studio/lib/interfaces/supportedLanguages";
+import {
+  DEFAULT_LANGUAGE_QUERY,
+  LANGUAGES_QUERY,
+} from "studio/lib/queries/languages";
+import {
+  SLUG_TRANSLATIONS_FROM_LANGUAGE_QUERY,
+  SLUG_TRANSLATIONS_TO_LANGUAGE_QUERY,
+} from "studio/lib/queries/slugTranslations";
+
+async function translateSlug(
+  slug: string,
+  targetLanguageId: string,
+  sourceLanguageId?: string,
+): Promise<string | undefined> {
+  if (slug.length === 0) {
+    return slug;
+  }
+  const slugTranslations = await client.fetch<SlugTranslations | null>(
+    sourceLanguageId !== undefined
+      ? SLUG_TRANSLATIONS_FROM_LANGUAGE_QUERY
+      : SLUG_TRANSLATIONS_TO_LANGUAGE_QUERY,
+    {
+      slug,
+      language: sourceLanguageId ?? targetLanguageId,
+    },
+  );
+  return slugTranslations?._translations.find(
+    (t) => t !== null && t.language === targetLanguageId,
+  )?.slug;
+}
+
+function negotiateClientLanguage(
+  availableLanguages: string[],
+): string | undefined {
+  const acceptLanguage = headers().get("Accept-Language");
+  if (acceptLanguage === null) return undefined;
+  return new Negotiator({
+    // Negotiator expects lower-case header name
+    headers: { "accept-language": acceptLanguage },
+  }).language(availableLanguages);
+}
+
+/**
+ * Handles rerouting of paths based on URL language code or user's preferred language
+ *
+ * Limitations/notes:
+ * - no special treatment of nested paths (e.g. `/blogg` uses slug `blogg`, while `/blogg/hei` uses slug `blogg/hei`)
+ * - given the following valid pages, an English user will be redirected from (1) to (2), and not to (3):
+ *      - (1) `/gift`
+ *      - (2) `/en/poison`
+ *      - (3) `/en/gift`
+ * - given the following valid pages, an English user navigating to (1) will not be redirected to (2), but receive a 404 error:
+ *      - (1) `/gift`
+ *      - (2) `/en/gift`
+ *
+ * @param request
+ */
+export async function languageMiddleware(
+  request: NextRequest,
+): Promise<NextResponse | undefined> {
+  const availableLanguages = await client.fetch<LanguageObject[] | null>(
+    LANGUAGES_QUERY,
+  );
+  if (availableLanguages === null) {
+    console.error("No languages available, language middleware aborted.");
```

Comment:
I'm not opposed to this, but perhaps we should find a better solution for error handling than adding logs to prod?

Reply:
Created #721 to note this